File: python/antwalk.py
from enum import Enum

class Direction(Enum):
    UP = 1
    DOWN = 2
    LEFT = 3
    RIGHT = 4

class Walk:
    def __init__(self, n):
        self.n = n
        self.world = ["."] * (n * n)
        self.world[0] = '*'
        self.ant = [0, 0]

    def has_fallen(self):
        return not all((0 <= c < self.n for c in self.ant))

    def walk(self, direction):
        if direction == Direction.UP:
            self.ant[1] -= 1
        elif direction == Direction.DOWN:
            self.ant[1] += 1
        elif direction == Direction.LEFT:
            self.ant[0] -= 1
        elif direction == Direction.RIGHT:
            self.ant[0] += 1
        fallen = self.has_fallen()
        if not fallen:
            self.world[self.ant[0] + (self.ant[1] * self.n)] = "*"
        return not fallen

    def __repr__(self):
        buf = ""
        for i in range(self.n):
            lptr =  i * self.n
            rptr =  (i+1) * self.n
            world_row = self.world[lptr:rptr]
            if i == self.ant[1] and not self.has_fallen():
                world_row[self.ant[0]] = 'x'
            buf += "".join(world_row) + "\n"
        return buf

def antwalk():
    print("Welcome to AntWalk!")
    walk = Walk(10)
    print(walk)
    while True:
        print("\n[U]p, [D]own, [L]eft, or [R]ight")
        try:
            command = input(">>> ").upper()
        except KeyboardInterrupt:
            return

        direction = None
        if command == "U":
            direction = Direction.UP
        elif command == "D":
            direction = Direction.DOWN
        elif command == "L":
            direction = Direction.LEFT
        elif command == "R":
            direction = Direction.RIGHT

        if direction is None:
            print("Command not recognised")
            continue

        result = walk.walk(direction)
        print(walk)

        if not result:
            print("Walk has concluded")
            print("Enter a filename to save your walk:")
            try:
                filename = input(">>> ")
            except KeyboardInterrupt:
                return
            try:
                with open(filename, "w") as outf:
                    print(walk, file=outf, end="")
            except (OSError, IOError) as file_err:
                print(file_err)
                print("Could not write walk to {}".format(filename))
            else:
                print("Walk successfully written to {}".format(filename))
            return

if __name__ == "__main__":
    antwalk()

File: rust/antwalk.rs
use std::fmt;
use std::fs::File;
use std::io;
use std::io::Write;

enum Direction {
    Up,
    Down,
    Left,
    Right,
}

struct Walk {
    n: usize,
    world: Vec<char>,
    ant: (usize, usize),
    fallen: bool,
}

impl Walk {
    fn new(n: usize) -> Walk {
        let mut walk = Walk {
            n: n,
            world: Vec::with_capacity(n * n),
            ant: (0, 0),
            fallen: false,
        };
        for _ in 0..(n * n) {
            walk.world.push('.');
        }
        walk.world[0] = '*';
        walk
    }

    fn walk(&mut self, direction: Direction) -> bool {
        let did_fall = match direction {
            Direction::Up => {
                if self.ant.1 == 0 {
                    true
                } else {
                    self.ant.1 -= 1;
                    false
                }
            }
            Direction::Down => {
                if self.ant.1 == (self.n - 1) {
                    true
                } else {
                    self.ant.1 += 1;
                    false
                }
            }
            Direction::Left => {
                if self.ant.0 == 0 {
                    true
                } else {
                    self.ant.0 -= 1;
                    false
                }
            }
            Direction::Right => {
                if self.ant.0 == (self.n - 1) {
                    true
                } else {
                    self.ant.0 += 1;
                    false
                }
            }
        };

        if did_fall {
            self.ant = (self.n + 1, self.n + 1);
            self.fallen = true;
        } else {
            self.world[self.ant.0 + (self.ant.1 * self.n)] = '*';
        }

        !did_fall
    }
}

impl fmt::Display for Walk {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut buf = String::new();
        for y in 0..self.n {
            for x in 0..self.n {
                if !self.fallen {
                    if (x, y) == self.ant {
                        buf.push('x');
                        continue;
                    }
                }
                buf.push(self.world[x + (y * self.n)]);
            }
            buf.push('\n');
        }
        write!(f, "{}", buf)
    }
}

fn antwalk() {
    println!("Welcome to AntWalk!");
    let mut walk = Walk::new(10);
    print!("{}", walk);

    loop {
        println!("[U]p, [D]own, [L]eft, or [R]ight");
        print!(">>> ");
        if io::stdout().flush().is_err() {
            println!("Could not flush stdin");
            return;
        }
        let mut command = String::new();
        match io::stdin().read_line(&mut command) {
            Ok(_) => (),
            Err(_) => {
                println!("Could not read stdin");
                return;
            }
        };

        let direction = match command.to_uppercase().trim() {
            "U" => Direction::Up,
            "D" => Direction::Down,
            "L" => Direction::Left,
            "R" => Direction::Right,
            _ => {
                println!("Command not recognised");
                continue;
            }
        };

        let result = walk.walk(direction);
        print!("{}", walk);

        if !result {
            println!("Walk has concluded");
            println!("Enter a filename to save your walk:");
            print!(">>> ");
            if io::stdout().flush().is_err() {
                println!("Could not flush stdin");
                return;
            }
            let mut filename = String::new();
            match io::stdin().read_line(&mut filename) {
                Ok(_) => (),
                Err(_) => {
                    println!("Could not read stdin");
                    return;
                }
            };

            let filename = filename.trim();

            let mut outf = match File::create(&filename) {
                Ok(f) => f,
                Err(_) => {
                    println!("Could not create file: {}", filename);
                    return;
                }
            };

            match outf.write_all(format!("{}", walk).as_bytes()) {
                Err(_) => {
                    println!("Could not write walk to {}", filename);
                }
                Ok(_) => {
                    println!("Walk successfully written to {}", filename);
                }
            }
            return;
        }
    }
}

fn main() {
    antwalk();
}
